### PR TITLE
Date and time fieldtypes overhaul

### LINKED ISSUE #3923: Date field gets rounded
## Bug Description
I've got a `date` field that gets populated from an external API. The date is saved as a timestamp. Whenever I save an entry the date will be rounded, which is very unfortunate.

```yaml
// Original value from the API
pledge_date: 1624875575
// 2021-06-28 10:19:35.0 +00:00

// Rounded value after save
pledge_date: 1624875540 
// 2021-06-28 10:19:00.0 +00:00
```

## How to Reproduce
1. Add a date field with `format: U` to your blueprint
2. Create an entry and add this value to the field: `1624875575`
3. Save the entry in the CP
4. The date will now be rounded to `1624875540`

## Environment
Statamic 3.1.26 Pro
Laravel 8.48.2
PHP 7.4.20
aerni/social-links 2.2.0
jonassiewertsen/statamic-livewire 2.5.0
visuellverstehen/statamic-classify 2.1.1

### PR DESCRIPTION
**Pull request by arthurperton**
Fixes #3923.

While trying to solve the issue, I also simplified the internal dataflow of the Vue components to make working on them easier. I ended up with solving and improving several things:

* Add support for (optionally) editing seconds
* Preserve seconds on save
* Simplified Vue component logic
* Upgrade v-calendar from v1 to v2
* Improve date range UI/UX
* Fixed `InvalidFormatException` when changing the `format` option, or config options that influence the default format e.g. `mode`, by falling back to Carbon::parse().

I also removed the separate `time_required` configuration option from the date field. Having a date field with time enabled but not required didn't seem very useful. Moreover the logic for guessing if a time was present or not in that case was (unavoidably) flawed and I saw no way of fixing it.

Please test and feel free to tweak the UI 👍🏻 

Screen Shot 2021-07-06 at 12 23 05

### CODE REVIEW

**Comment by arthurperton** [comment]
Not sure what is happening, the tests run fine locally..

File: package.json [hunk 1]
```diff
@@ -51,7 +51,7 @@
     "tiptap-extensions": "^1.28.6",
     "underscore": "~1.9.2",
     "uniqid": "^5.2.0",
-    "v-calendar": "^1.0.1",
+    "v-calendar": "^2.3.0",
```
Comment (jasonvarga) [inline package.json:54]: Bumping a major version might be an issue for people using v-calendar themselves. This should not happen in a 3.1 release. Let's target `master`.

**Review by jasonvarga — CHANGES_REQUESTED** [review]
> I also removed the separate time_required configuration option from the date field. Having a date field with time enabled but not required didn't seem very useful. 

I assume we added it because people wanted it. (Maybe not, though.)

I can see the use case of two different date fields, one with time and one without.
But I also don't see the point in having one date field where some entries have time and some don't.

I don't think we can just rip it out though. At least not in 3.1.

Let's target `master` so this gets into 3.2 instead and we can call it a breaking change.

If people were actually using this feature... I don't have a suggestion for them though. Do we just hope no one was using it?

@jackmcdade ?

**Comment by jasonvarga** [comment]
Great PR though!

File: package.json (hunk 1 above)
Comment (arthurperton) [reply package.json:54]: Will do. Let's keep an eye open then, they've currently even have a v3 in alpha.

**Comment by arthurperton** [comment]
> Great PR though!

Glad you like it 😄

**Comment by arthurperton** [comment]
I just merged the master branch into this PR. The unit tests failed again. And I got Duncan laughing at me 🤣 

It is just `Tests\Feature\GraphQL\Fieldtypes\DateFieldtypeTest::it_gets_dates` that fails, which again runs fine locally. Any tips @jasonvarga?

**Comment by jackmcdade** [comment]
FYO I made some updates to the `Date` Fieldtype in 3.2 as part of the `ui-improvements` branch: https://github.com/statamic/cms/commit/76a88d5c55d4c0bd5ab6b6120524d176c52fa6fa

As for the time option/required stuff, yeah it's a bit hairy. MAYBE no one was using it, and it's _mostly_ a UI thing, but not quite. 🤔

**Comment by arthurperton** [comment]
I hope `time_required` can be left out, now that we're targeting 3.2. Support for it _can_ be put back in if necessary, but it will always be kind of half-baked as far as I can see.

**Comment by jasonvarga** [comment]
The test fails locally if you're using the same requirements as that test image.
```
composer require "laravel/framework:^6.6.2" "orchestra/testbench:4.*"  --no-interaction --no-update
composer update --prefer-lowest --prefer-dist --no-interaction --no-suggest
```

(Put `config(['app.debug' => true]);` at the start of the test to see a better error)

**Comment by arthurperton** [comment]
> FYO I made some updates to the `Date` Fieldtype in 3.2 as part of the `ui-improvements` branch: [76a88d5](https://github.com/statamic/cms/commit/76a88d5c55d4c0bd5ab6b6120524d176c52fa6fa)

@jackmcdade Do you want me to bring over the button classes into this PR or ...?

**Comment by jackmcdade** [comment]
I suppose we can work it out on the merge. No worries.

**Comment by arthurperton** [comment]
Thanks Jason, that did the trick ✨

**Comment by jasonvarga** [comment]
> Do you want me to bring over the button classes into this PR or ...?

> I suppose we can work it out on the merge. No worries.

Actually, now that the ui branch has been merged into master - would you mind resolving the conflicts here? 🙂

**Comment by jackmcdade** [comment]
@arthurperton Are you able to wrap this one up?

**Comment by arthurperton** [comment]
Hey Jack! Thanks for reminding me. I was just wondering this morning this didn’t make it into 3.2 yet haha. Wasn’t aware (anymore) that there was still some fixing to do here. Putting it on top of my list now.

**Comment by jasonvarga** [comment]
Don't forget to target `3.2` now 👍

**Comment by arthurperton** [comment]
This one just missed the boat for 3.2. Shall we integrate this into 3.3? 

If so I will target the `master` branch and resolve any conflicts, so that it is ready for merging.

**Comment by jasonvarga** [comment]
Yes, let's try to get it into 3.3.

File: package.json (hunk 1 above)
Comment (jackmcdade) [reply package.json:59]: I'm fine with removing `time_required` – it was never request, I think I was just overaggressively wiring up props. If you have a time field, it'll get _a_ time, and therefore be filled. Seems superfluous.

**Review by jackmcdade — APPROVED** [review]

**Pull request merged**